File: src/main/java/dark/leech/text/models/Properties.java
package dark.leech.text.models;

import java.util.List;

import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
public class Properties {
    private String name; // Tên truyện
    private String author; // Tác giả
    private String url; // Đường dẩn truyện
    private String cover; // Đường dẫn Cover
    private List<Chapter> chapList; // Danh sách chương
    private List<Pager> pageList; // Danh sách trang
    private boolean forum; // Trang get có phải forum hay không
    private int size; // Số chương
    private String savePath; // Thư mục lưu
    private String gioiThieu;
    private boolean addGt;
    private String charset = "UTF-8";
    private String[] urlList;

    public void setCover(String cover) {
        this.cover = cover;
    }

    public void setCover(String cover, String page) {
        if (cover == null) return;
        if (cover.startsWith("http")) this.cover = cover;
        else this.cover = page + cover;
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/models/Properties.kt
package dev.haipham22.leechtext.models

/**
 * properties.json của mỗi book (port từ models/Properties.java).
 * Không phải java.util.Properties — model per-book.
 */
data class Properties(
    var name: String? = null,
    var author: String? = null,
    var url: String? = null,
    var cover: String? = null,
    var chapList: List<Chapter>? = null,
    var pageList: List<Pager>? = null,
    var forum: Boolean = false,
    var size: Int = 0,
    var savePath: String? = null,
    var introduce: String? = null,
    var addGt: Boolean = false,
    var charset: String? = "UTF-8",
    var urlList: Array<String>? = null,
    /** True = Đang ra, false = Hoàn thành — từ field `ongoing` của detail.js. */
    var ongoing: Boolean? = null,
) {
    /** Resolve cover tương đối theo URL trang. */
    fun setCover(
        cover: String?,
        page: String?,
    ) {
        if (cover == null) return
        this.cover = if (cover.startsWith("http")) cover else page + cover
    }

    override fun equals(other: Any?): Boolean = other is Properties && name == other.name && author == other.author &&
        url == other.url && cover == other.cover && chapList == other.chapList &&
        pageList == other.pageList && forum == other.forum && size == other.size &&
        savePath == other.savePath && introduce == other.introduce &&
        addGt == other.addGt && charset == other.charset &&
        urlList.contentEquals(other.urlList)

    override fun hashCode(): Int = urlList?.contentHashCode() ?: 0
}
